[Data] Fix ActorPool autoscaler to properly scale up

alexeykudinkin opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

Currently, ActorPool autoscaling actually isn't scaling up until all of its `num_actors * max_tasks_in_flight` slots are full. This is a seriously limiting factor:

 - It requires AP to accumulate backlog of 4x its # of actors before autoscaling even starts
 - 3 out of 4 tasks are sitting in the queue, meaning their execution won't start until previous ones complete

Changes
---

1. Revisiting Actor Pool autoscaling protocol to be based on utilization defined as `# of submitted tasks / number of running actors * max_concurrency` 
2. Added `AutoscalingConfig` to make all configuration explicitly available inside `DataContext`
3. Setting default upscaling threshold to 200% (currently is set to 400%)
4. Updated tests
5. Added docs


## Related issue number

<!-- For example: "Closes #1234" -->

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a
           method in Tune, I've added it in `doc/source/tune/api/` under the
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

alexeykudinkin committed:
Tidying up

Signed-off-by: Alexey Kudinkin <ak@anyscale.com>

alexeykudinkin committed:
Relocated AP util to AP itself

Signed-off-by: Alexey Kudinkin <ak@anyscale.com>

alexeykudinkin committed:
Reduce default # of max tasks in-flight;
Tidying up

Signed-off-by: Alexey Kudinkin <ak@anyscale.com>

alexeykudinkin committed:
Wire in actor memory reqs properly

Signed-off-by: Alexey Kudinkin <ak@anyscale.com>

alexeykudinkin committed:
Increase AP autoscaling util threshold to 95%;
Ignore free slots when autoscaling

Signed-off-by: Alexey Kudinkin <ak@anyscale.com>

alexeykudinkin committed:
Cleaning up tests

Signed-off-by: Alexey Kudinkin <ak@anyscale.com>

alexeykudinkin committed:
Introduced `_AutoscalingAction`

Signed-off-by: Alexey Kudinkin <ak@anyscale.com>

alexeykudinkin committed:
Replaced `scale_up`/`scale_down` API with new `apply`

Signed-off-by: Alexey Kudinkin <ak@anyscale.com>

alexeykudinkin committed:
Make sure AP doesn't scale up/down past limits

Signed-off-by: Alexey Kudinkin <ak@anyscale.com>

alexeykudinkin committed:
Added appropriate deltas to scaling actions

Signed-off-by: Alexey Kudinkin <ak@anyscale.com>

alexeykudinkin committed:
Fixed tests

Signed-off-by: Alexey Kudinkin <ak@anyscale.com>

alexeykudinkin committed:
Tidying up

Signed-off-by: Alexey Kudinkin <ak@anyscale.com>

alexeykudinkin committed:
Fixed circular imports

Signed-off-by: Alexey Kudinkin <ak@anyscale.com>

alexeykudinkin committed:
Retired `_AutoscalingActionKind`, replaced with delta sign encoding;

Signed-off-by: Alexey Kudinkin <ak@anyscale.com>

alexeykudinkin committed:
Updated tests

Signed-off-by: Alexey Kudinkin <ak@anyscale.com>

alexeykudinkin committed:
`lint`

Signed-off-by: Alexey Kudinkin <ak@anyscale.com>

alexeykudinkin committed:
Tidying up

Signed-off-by: Alexey Kudinkin <ak@anyscale.com>

alexeykudinkin committed:
Fixing usages

Signed-off-by: Alexey Kudinkin <ak@anyscale.com>

alexeykudinkin committed:
Tidying up more

Signed-off-by: Alexey Kudinkin <ak@anyscale.com>

alexeykudinkin committed:
Fixing remaining usages

Signed-off-by: Alexey Kudinkin <ak@anyscale.com>

alexeykudinkin committed:
`lint`

Signed-off-by: Alexey Kudinkin <ak@anyscale.com>

alexeykudinkin committed:
Tidiyng up more

Signed-off-by: Alexey Kudinkin <ak@anyscale.com>

alexeykudinkin committed:
Fixing more tests

Signed-off-by: Alexey Kudinkin <ak@anyscale.com>

alexeykudinkin committed:
Fixed refs

Signed-off-by: Alexey Kudinkin <ak@anyscale.com>

alexeykudinkin committed:
Abstracted dedicated `AutoscalingConfig`

Signed-off-by: Alexey Kudinkin <ak@anyscale.com>

alexeykudinkin committed:
Added `AutoscalingConfig` to `DataContext`

Signed-off-by: Alexey Kudinkin <ak@anyscale.com>

alexeykudinkin committed:
Fixed tests

Signed-off-by: Alexey Kudinkin <ak@anyscale.com>

alexeykudinkin committed:
Removed dependency on internal RD constant

Signed-off-by: Alexey Kudinkin <ak@anyscale.com>

alexeykudinkin committed:
Added `DataContext.max_per_actor_task_queue_size`

Signed-off-by: Alexey Kudinkin <ak@anyscale.com>

alexeykudinkin committed:
Rebased `max_tasks_in_flight_per_actor` to leverage `DataContext.max_per_actor_task_queue_size`

Signed-off-by: Alexey Kudinkin <ak@anyscale.com>

alexeykudinkin committed:
`max_per_actor_task_size` > `max_tasks_in_flight_per_actor`;
Made default to be 2 x of `max_concurrency`

Signed-off-by: Alexey Kudinkin <ak@anyscale.com>

alexeykudinkin committed:
`lint`

Signed-off-by: Alexey Kudinkin <ak@anyscale.com>

alexeykudinkin committed:
Missing import

Signed-off-by: Alexey Kudinkin <ak@anyscale.com>

alexeykudinkin committed:
`lint`

Signed-off-by: Alexey Kudinkin <ak@anyscale.com>

alexeykudinkin committed:
Tidying up

Signed-off-by: Alexey Kudinkin <ak@anyscale.com>

alexeykudinkin committed:
Rebased AP util to be expressed as # of submitted tasks / max running tasks

Signed-off-by: Alexey Kudinkin <ak@anyscale.com>

alexeykudinkin committed:
Updated py-doc

Signed-off-by: Alexey Kudinkin <ak@anyscale.com>

alexeykudinkin committed:
Bumped autoscaling threshold to 100%

Signed-off-by: Alexey Kudinkin <ak@anyscale.com>

alexeykudinkin committed:
Fixed circular imports

Signed-off-by: Alexey Kudinkin <ak@anyscale.com>

alexeykudinkin committed:
Relocated `AutoscalingConfig` under ray.data.context

Signed-off-by: Alexey Kudinkin <ak@anyscale.com>

alexeykudinkin committed:
`lint`

Signed-off-by: Alexey Kudinkin <ak@anyscale.com>

alexeykudinkin committed:
Fixing imports

Signed-off-by: Alexey Kudinkin <ak@anyscale.com>

alexeykudinkin committed:
Fixing autoscaling_config default

Signed-off-by: Alexey Kudinkin <ak@anyscale.com>

alexeykudinkin committed:
Fixed _ActorPool creation

Signed-off-by: Alexey Kudinkin <ak@anyscale.com>

alexeykudinkin committed:
Fixing invalid field set up

Signed-off-by: Alexey Kudinkin <ak@anyscale.com>

alexeykudinkin committed:
Updated test fixtures

Signed-off-by: Alexey Kudinkin <ak@anyscale.com>

alexeykudinkin committed:
Updated more fixtures

Signed-off-by: Alexey Kudinkin <ak@anyscale.com>

alexeykudinkin committed:
`lint`

Signed-off-by: Alexey Kudinkin <ak@anyscale.com>

alexeykudinkin committed:
Tidying up

Signed-off-by: Alexey Kudinkin <ak@anyscale.com>

lk-chen commented on python/ray/llm/_internal/batch/processor/sglang_engine_proc.py:
```diff
@@ -6,9 +6,6 @@
 from pydantic import Field, root_validator
 
 import ray
-from ray.data._internal.execution.operators.actor_pool_map_operator import (
```
Don't want people get confused where the magic number 4 comes from. Could you add following to top of these 2 files

```
# Higher values here are better for prefetching and locality. It's ok for this to be
# fairly high since streaming backpressure prevents us from overloading actors.
DEFAULT_MAX_TASKS_IN_FLIGHT = 4
```

alexeykudinkin committed:
Added a validation of autoscaling config to AP

Signed-off-by: Alexey Kudinkin <ak@anyscale.com>

alexeykudinkin committed:
Added reason for scaling down rejection

Signed-off-by: Alexey Kudinkin <ak@anyscale.com>

alexeykudinkin committed:
Don't scale down if not allowed

Signed-off-by: Alexey Kudinkin <ak@anyscale.com>

alexeykudinkin committed:
Check whether allowed to scale down

Signed-off-by: Alexey Kudinkin <ak@anyscale.com>

raulchen reviewed (commented):
overall LGTM

alexeykudinkin committed:
Revisited scaling protocol to
 - Make validation whether request can be applied private (removed dup)
 - Log debugging warnings internally

Signed-off-by: Alexey Kudinkin <ak@anyscale.com>

alexeykudinkin committed:
Fixed tests

Signed-off-by: Alexey Kudinkin <ak@anyscale.com>

alexeykudinkin committed:
Fixing more tests

Signed-off-by: Alexey Kudinkin <ak@anyscale.com>

alexeykudinkin committed:
`lint`

Signed-off-by: Alexey Kudinkin <ak@anyscale.com>

alexeykudinkin committed:
Updated test

Signed-off-by: Alexey Kudinkin <ak@anyscale.com>

alexeykudinkin commented on python/ray/data/_internal/execution/operators/actor_pool_map_operator.py:
```diff
@@ -132,15 +131,30 @@ def __init__(
+                #       running task we'd allow 1 more task to be enqueued
+                compute_strategy.max_tasks_in_flight_per_actor
+                or data_context.max_tasks_in_flight_per_actor
+                or max_actor_concurrency * 2
```
Captured it in the comment above, not really sure what value the constant will add

bveeramani replied:
+1 to defining constant. It's easier to understand if you're skimming the code and not reading all of the comments

alexeykudinkin commented on python/ray/data/_internal/execution/operators/actor_pool_map_operator.py:
```diff
@@ -758,20 +777,50 @@ def can_scale_down(self):
+            # Make sure after scaling down actor pool size won't fall below its
+            # min size
+            target_num_actors = min(
+                abs(config.delta), max(self.current_size() - self.min_size(), 0)
```
Fair point. I was actually thinking to just shutdown the actor pool once processing is done, but now that i'm thinking about it, it's better to incrementally release workers as soon as they could be released.

alexeykudinkin commented on python/ray/data/context.py:
```diff
@@ -338,6 +380,11 @@ class DataContext:
             actors to start before raising a timeout, in seconds.
+        max_tasks_in_flight_per_actor: Max number of tasks that could be enqueued
+            into the individual Actor's queue. Note that running tasks are not counted
+            towards this limit. This setting allows Actors to start pulling and buffering
```
Yeah, this is remnant of when this parameter was not overlapping with max_concurrency. Will update

alexeykudinkin committed:
Tidying up

Signed-off-by: Alexey Kudinkin <ak@anyscale.com>

alexeykudinkin commented on python/ray/data/_internal/execution/autoscaler/autoscaling_actor_pool.py:
```diff
@@ -1,10 +1,32 @@
 
 
+@dataclass
+class ScalingConfig:
```
Yeah, i've realized that this is rather an intent/request

alexeykudinkin commented on python/ray/data/_internal/execution/operators/actor_pool_map_operator.py:
```diff
@@ -758,20 +777,50 @@ def can_scale_down(self):
+                config.delta, max(self.max_size() - self.current_size(), 0)
+            )
+
+            logger.info(
```
SG

alexeykudinkin committed:
Tidying up some more

Signed-off-by: Alexey Kudinkin <ak@anyscale.com>

alexeykudinkin committed:
Avoid enforcing scaling min/max inside AP itself to allow it to be scaled down to 0 (upon completion)

Signed-off-by: Alexey Kudinkin <ak@anyscale.com>

alexeykudinkin committed:
Added test to verify we're able to scale down to 0

Signed-off-by: Alexey Kudinkin <ak@anyscale.com>

alexeykudinkin committed:
`lint`

Signed-off-by: Alexey Kudinkin <ak@anyscale.com>

alexeykudinkin committed:
Updated py-doc

Signed-off-by: Alexey Kudinkin <ak@anyscale.com>

alexeykudinkin committed:
Fixed py-docs

Signed-off-by: Alexey Kudinkin <ak@anyscale.com>

alexeykudinkin committed:
Tidying up

Signed-off-by: Alexey Kudinkin <ak@anyscale.com>

alexeykudinkin committed:
Fixing more py-docs

Signed-off-by: Alexey Kudinkin <ak@anyscale.com>

alexeykudinkin committed:
Fixed more tests

Signed-off-by: Alexey Kudinkin <ak@anyscale.com>

bveeramani commented on python/ray/data/_internal/execution/autoscaler/autoscaling_actor_pool.py:
```diff
@@ -61,46 +89,20 @@ def num_free_task_slots(self) -> int:
     @abstractmethod
-    def per_actor_resource_usage(self) -> ExecutionResources:
-        """Per actor resource usage."""
+    def get_pool_util(self) -> float:
+        """Calculate the utilization of the given actor pool."""
         ...
```
Should this be the responsibility of the autoscaler? You can compute this just using public actor pool methods, and I think the actor pool interface is already really bloated

alexeykudinkin replied:
Yeah, i went back and forth on this. Ultimately decided that the scope of AP should include defining what util actually means (in terms of # of actors used, concurrency slots or what not) as these are intrinsic actor details. Autoscaler on the other hand operates at a higher level (if util > threshold then upscale for ex)

bveeramani reviewed (approved):
LGTM.

@alexeykudinkin I'm running the release tests here: https://buildkite.com/ray-project/release/builds/46720#.

Would you mind skimming it and making sure there's no major regression before merging?

alexeykudinkin committed:
Fixed baseline doc

Signed-off-by: Alexey Kudinkin <ak@anyscale.com>

alexeykudinkin committed:
Restore default of max-tasks in-flight to 4

Signed-off-by: Alexey Kudinkin <ak@anyscale.com>

alexeykudinkin commented on python/ray/data/_internal/execution/operators/actor_pool_map_operator.py:
```diff
@@ -736,42 +769,84 @@ def num_pending_actors(self) -> int:
+            return -num_released
 
-        return num_actors
+        return None
```
Just more clearly signal that there's no action taken (rather than attempted and failed for ex in case of downscaling)

alexeykudinkin commented on python/ray/data/_internal/execution/operators/actor_pool_map_operator.py:
```diff
@@ -736,42 +769,84 @@ def num_pending_actors(self) -> int:
+            return -num_released
 
-        return num_actors
+        return None
```
Answered this comment from Hao above (None to signal no action taken)

alexeykudinkin committed:
Tidying up

Signed-off-by: Alexey Kudinkin <ak@anyscale.com>
